### ISSUE TITLE
PKCS#11 workflow with ALWAYS_AUTHENTICATE private keys

### ISSUE THREAD

**Issue by Jakuje** [OpenSC/OpenSC#1269]
### Problem Description
The "correct" workflow of PKCS#11 operations involving ALWAYS_AUTHENTICATE key is following:
```
C_Login()
C_FindObjectsInit()
C_FindObjects()
C_FindObjectsFinal()
C_GetAttributeValue(CKA_ALWAYS_AUTHENTICATE)
C_SignInit()
C_Login(CKU_CONTEXT_SPECIFIC)
C_Sign()
```
This means that for one-shot applications, user needs to write a PIN twice.

There is a chicken-egg problem with `ALWAYS_AUTHENTICATE` attribute of private keys:
 * One can not start the `C_SignInit` operation before the `C_Login` was performed, object handle obtained and the `CKA_ALWAYS_AUTHENTICATE` attribute detected.
 * One can not use `C_Login()` before `C_SignInit()`, because after #1256, the card is losing login state (using keepalive messages).

In OpenSC 0.17.0, also the workflow without the second login worked for legacy applications (for the first signature), but it is not true after merging the #1256:
```
C_Login()
C_FindObjectsInit()
C_FindObjects()
C_FindObjectsFinal()
C_SignInit()
C_Sign()
```

### Proposed Resolution
There  is no good solution. There are several workarounds, such as PIN caching, ignoring this restriction (`pin_cache_ignore_user_consent = true`).

More complicated solutions might be introducing some timers (might interfere with the calling application). For more options, see the [dengert's comment](https://github.com/OpenSC/OpenSC/pull/1256#issuecomment-366073468). I am inclining to solution in this direction, unless it will be too complicated and fragile.

We should not obtain the card log in the `C_Login()` infinitely, because for example `ssh-agent` workflow does not immediately follow the `C_SignInit` operation.

### Steps to reproduce
Simplest use case is one-shot OpenSSH session with yubikey and the "digital signature key" used for ssh authentication:
```
ssh-keygen -D /path/to/opensc-pkcs11.so >> ~/.ssh/authorized_keys ## select only the signature key!
ssh -I /path/to/opensc-pkcs11.so localhost
```

### Logs
Working with OpenSC 0.17.0 (from #1256): https://github.com/OpenSC/OpenSC/files/1736542/opensc-debug.log

**Comment by mouse07410** [comment]
I'm not sure at all we want to expend much efforts and increase fragility of the code, merely to accommodate the workflow that's wrong to begin with.

And there already is a solution: define `pin_cache_ignore_user_consent = true;` in `opensc.conf`. IMHO it's better to reduce security of one installation by setting that config, than to increase risks to all of the installation by making the code less robust or less flexible.

**Comment by dengert** [comment]
It would appear that trying to support the legacy applications at the same time trying to support multiple applications on a card and interference from other processes can not be solved, at least for a real PIV type card without some pin caching. 

The Pin caching could also be done in OpenSSH. 

In the PIV case, the C_Login CKU_USER and CKU_CONTEXT_SPECIFIC  use the same PIN. It is not  clear if that is a PKCS#11 requirement or not. In which case PIN caching would not work. 

So do we need a opensc.conf option to not do the #1256, which is in line with the current code?

I feel the "Keep_alive" in #1243 is backwards, The user has to specify by ATR if they want it or not. 
I would argue that we should be doing our best to prevent interference so "keep_alive" would be on be default, even though it adds some extra overhead, to make things work in most cases. 

The PIV card enforced policy on the card  for the Sign Key is the issue we are facing. Would a legacy option be a better way  to turn off #1256?

**Comment by frankmorgner** [comment]
The card flags in `opensc.conf` are a hint from the user to tell OpenSC what to do. Nothing prevents us from enabling `keep_alive` when detecting multiple applets on the card (e.g by iterating through all available driver's `match_card()` without calling `init()` afterwards); @dengert , you may even activate the flag if you detect this in your card driver. However, due to the side effects, I'd opt for keeping this disabled without user consent.


To extend @mouse07410 's [comment](https://github.com/OpenSC/OpenSC/issues/1269#issuecomment-366720703), we could implement application specific configuration blocks. We are already distinguishing between tokend, minidriver, pkcs11 and default. We could extend this system to bind some options to a name of a process (e.g. *`ssh`*). A poor man's implementation of this idea can be achieved by passing `OPENSC_CONF`:
```
OPENSC_CONF="/path/to/special/opensc.conf" ssh -I /path/to/opensc-pkcs11.so localhost
```


An other possibility would be to adapt the implementation of the PIV driver so that there are no other APDU sent between login and signature. If I'm not mistaken, there is theoretically no need to send any APDU in the following library calls:
```
C_FindObjectsInit()
C_FindObjects()
C_FindObjectsFinal()
C_GetAttributeValue(CKA_ALWAYS_AUTHENTICATE)
C_SignInit()
C_Login(CKU_CONTEXT_SPECIFIC)
```
`C_Login(CKU_CONTEXT_SPECIFIC)` disables the keep-alive command and the signature is performed directly after the VERIFY.

**Comment by frankmorgner** [comment]
@dengert and @Jakuje what's the status of this issue, has it been fixed with #1307 ?

**Comment by dengert** [comment]
I can respond tonight or tomorrow morning.

**Comment by dengert** [comment]
No #1307 does not solve this issue. That dealt with trying not to cause a reset of the login_state.
 
The comment about only calling a card's match_card will not work because a number of card drivers do   a SELECT FILE/AID command for their applet  and many cards will reset their login state, even though the specs say this should not happen. 
  
This issue is a direct consequence of the PKCS\#11 specifications, combined with what some cards can do to enforce a policy of requiring the PIN to be verified before a crypto operation. PIV, OpenPGP and Belpic. There may be others. 

PKCS#11 does not allow access to private keys without a C_Login() first. To avoid the C_Login, an application could test the login state before doing a C_Login. This would only work if (PINS are cached) or disconnect=leave was used by all applications and we have a single applet per card. Trying bounce back and forth between OpenPGP and PIV applets o the same device resets the login state of the other applet. The device vendor would need to get involved for example using the Global PIN. 

A key that requires CKA_ALWAYS_AUTHENTICATE and enforced by the card is there to implement Policy rules by the issuer of the card. We should not be using that keys. 
  
Thus the application must do a  C_Login(CKU_CONTEXT_SPECIFIC) before the C_Sign or C_SignFinal. 

The known cards that enforce some restrictions on VERIFY and crypto operations for some keys are:

- PIV requires a VERIFY PIN be the command immediately before a crypt command using the 9C  Sign key. The same PIN and VERIFY command is used C_Login() and C_Login(CKU_CONTEXT_SPECIFIC)

- OpenPGP  VERIFY says: " PW 1 with P2 = 81 sets the access conditions for a PSO:CDS command only. Depending on the PW1 status byte (see Extended capabilities) this access condition is only valid for one PSO:CDS command or remains valid for several attempts." and in PSO: COMPUTE DIGITAL SIGNATURE it says: The command is only possible after correct presentation of PW1 with No. 81. The command internally checks the PW1 Status byte DO (first byte), if the value is 00, then the access condition (No. 81 of PW1) is reset and the PW has to be verified again for the following command. Same PIN but different VERIFY commands. One way to implement this is is to use CKU_CONTEXT_SPECIFIC. No restriction of other commands between the the VERIFY and PSO.

- Belpic  has the same restriction as PIV. See #1295 and  https://github.com/OpenSC/OpenSC/issues/1295#issuecomment-379351202 But the driver has not been modified yet handle the problem of the VERIFY preceding the  Select Key File and Set Security Env commands. 

Commit 27add2ee3c10269fba9465616a49231cfb68c65f was designed to inform a driver C_Login(CKU_CONTEXT_SPECIFIC)  is requesting a VERIFY PIN. card-piv.c was modified to use this and force the VERIFY and the next card command to be in the same PCSC Transaction, thus avoiding any interference caused by other drivers in this process or other processors from being issued to the card between the VERIFY and Crypto operation. 

**SO WHAT IS THIS ISSUE IS ABOUT?** 

It is about the user having  to entry the PIN only once. With respect to the PIV card, if all the certificates are read before the first VERIFY and the next card command is the crypto SIgn operation, the the VERIFY sets the logged_in state to true and allows the SIGN. So no caching of the pin is required. But interference can still occur, and there is no time limit on how long the time can be between these two commands. The flow of PKCS#11 commands gets in the way too. 
 
But it looks like to me, a single VERIFY will not work with  OpenPGP because you need to do one VERIFY to login to the card, then read a number of DO's then using the same PIN but with VERIFY with a different P2.

The Belpic has similar issues. 

**SO WHAT CAN WE DO**

See @Jakuje comments above. 

I would also add don't use a key that requires CKA_ALWAYS_AUTHENTICATE for authentication, but only for digital signatures. OpenPGP and PIV and Belpic use CKA_ALWAYS_AUTHENTICATE for Policy reasons.

**Comment by Jakuje** [comment]
I believe everything was already said and explained by @dengert. I am really not sure if there is a solution for this problem. This issue was filled mostly for the tracking that there is a change for specific (mis)configurations we should be aware of.

I don't like much the idea of caching the pin (by not calling the VERIFY command on `C_Login`), when the pin caching is not explicitly configured. In some cards, this might result in unexpected results. But for example in PIV, I believe the structure is quite well known and the login state can be "emulated" to provide this functionality. What I am a bit afraid is the security of storing the pin little bit longer than necessary.

**Comment by mouse07410** [comment]
I don't think there is a good solution.

IMHO there are two workarounds:
- cache PIN and use ignore_consent config option - which is already available; or
- check return code of C_Sign () and if it's USER_NOT_LOGGED_IN - call C_Login and re-issue C_Sign. This works for me in libp11 (my fork).

**Comment by mouse07410** [comment]
@frankmorgner I think the correct way for an *application* to respect ALWAYS_AUTHENTICATE is to invoke C_Login and re-issue the command (C_Sign or C_Decrypt correspondingly) if the return code from the previous command was USER_NOT_LOGGED_IN.
Explicitly forcing C_Login before, e.g., C_Sign interferes with the normal workflow, as I experienced with the upstream of libp11 (and it's one of the reasons why I'm maintaining my fork of it).

**Comment by Jakuje** [comment]
@mouse07410 Your second workaround will not help in case you want to call `C_Login()` only once (and prompt a user for a PIN only once), as it worked before.

@mouse07410 no, that was the old way how it was done before the ALWAYS_AUTHENTICATE was somehow standardized, if I remember well. But anyway, it does not help in in any case, that if the private key is hidden before login, you can NOT obtain its handle without doing `C_Login()` beforehead.

The correct way is really to check the `ALWAYS_AUTHENTICATE` flag and then do the login or not in the correct place as said by the [PKCS#11 specs (4.9 Private key objects):](http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html#_Toc416959719)

> Re-authentication occurs by calling C_Login with userType set to CKU_CONTEXT_SPECIFIC immediately after a cryptographic operation using the key has been initiated (e.g. after C_SignInit).

**Comment by mouse07410** [comment]
What I want is to call C_Login only when the card demands it, not when the middleware thinks it's the right time to do so. The workflow I described did it for me. I don't know how OpenSSH differs from pkcs11-tool or libp11, so it works for the later but didn't work for OpenSSH...

**Comment by dengert** [comment]
I believe the primary application we are addressing here is OpenSSH and agent? And then only if it is trying to use a key that has ALWAYS_AUTHENTICATE and enforced by the card.

I would suggest that we do nothing in the code, but recommend, based on the card and policies of the card issuer that this key should not be used in this manor. 

Some cards may have a workaround. Use other keys on the card, or turn off the card's enforcing of the ALWAYS_AUTHENTICATE. 
For example: PIV specs define 20 extra retired encryption keys and certificates The History object defines how many there are, and if the certificates are on or off the card.  There is code in card-piv.c to allow these to be used for any use based on the keyUsage in the certificate, but only if this is not a US gov issued card. This requires a CHUID with the FASC-N with the agency code of 9999 which indicates it is not a US gov issued card.

The Yubico piv-tool can create such a CHUID (really needs to be checked, I have not looked closely at it.)
I don't see a create a History Object command with their tool yet. The history object can be create using piv-tool -s "DER of the history object" Their tool can import or generate keys on their devices.  

Can someone who uses a Yubikey clarify if then can set or turn off the card enforcement of ALWAYS_AUTHENTICATE? We could then change pkcs15-piv.c  to test this before doing 
`prkey_obj.user_consent = prkeys[i].user_consent;` for the sign key. 

Yubico should be involved in this, and add the needed commands and documentyation to their yubico-piv tool. 

**In other words** may it easier for a user to not  misuse a ALWAYS_AUTHENTICATE key.

**Comment by frankmorgner** [comment]
If there's not much to do here, then this should be documented in the wiki rather than as issue. All the background and workarounds can be written down there.
